lib/ansible/modules/cloud/amazon/ec2_vpc_igw.py

```diff
@@ -80,128 +83,205 @@
+        tags = self._module.params.get('tags')
 
+        result = None
         try:
-            igw = vpc_conn.create_internet_gateway()
-            vpc_conn.attach_internet_gateway(igw.id, vpc_id)
-            changed = True
-        except EC2ResponseError as e:
+            if state == 'present':
+                result = self.ensure_igw_present(vpc_id, tags)
+            elif state == 'absent':
+                result = self.ensure_igw_absent(vpc_id)
+        except AnsibleIGWException as e:
+            self.fail(msg=str(e))
+
+    def get_matching_igw(self, vpc_id):
+        filters = ansible_dict_to_boto3_filter_list({'attachment.vpc-id': vpc_id})
+        igws = None
+        try:
+            response = self._connection.describe_internet_gateways(Filters=filters)
+            igws = response['InternetGateways']
+        except botocore.exceptions.ClientError as e:
+            self.fail(msg=str(e))
+
+
+
+        igw = None
+        if len(igws) > 1:
             raise AnsibleIGWException(
-                'Unable to create Internet Gateway, error: {0}'.format(e))
+                'EC2 returned more than one Internet Gateway for VPC {0}, aborting'.format(vpc_id))
+        elif igws:
+            igw = camel_dict_to_snake_dict(igws[0])
+
+        return igw
+
+    def check_input_tags(self, tags):
+        nonstring_tags = [k for k, v in tags.items() if not isinstance(v, string_types)]
+        if nonstring_tags:
+            self.fail(msg='One or more tags contain non-string values: {0}'.format(nonstring_tags))
+
+    def ensure_tags(self, igw_id, tags, add_only):
+        final_tags = []
+
+        filters = ansible_dict_to_boto3_filter_list({'resource-id': igw_id, 'resource-type': 'internet-gateway'})
+        try:
+            cur_tags = self._connection.describe_tags(Filters=filters)
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            self._module.fail_json_aws(e, msg="Couldn't describe tags")
+
+        to_update, to_delete = compare_aws_tags(boto3_tag_list_to_ansible_dict(cur_tags.get('Tags')), tags, True)
+        final_tags = boto3_tag_list_to_ansible_dict(cur_tags.get('Tags'))
+
+        if to_update:
+            try:
+                if self._check_mode:
+                    # update tags
+                    final_tags.update(to_update)
+                else:
+                    AWSRetry.exponential_backoff(
+                        catch_extra_error_codes=['InvalidInternetID.NotFound']
+                    )(self._connection.create_tags)(
+                        Resources=[igw_id],
+                        Tags=ansible_dict_to_boto3_tag_list(to_update)
+                    )
+
+                self._results['changed'] = True
+            except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+                self._module.fail_json_aws(e, msg="Couldn't create tags")
 
-    igw.vpc_id = vpc_id
+        if to_delete and not add_only:
```

Comment:
I think rather than checking for to_delete and add_only here (and having two hardcoded values for add_only and the purge_tags parameter for compare_aws_tags), I'd consolidate the options since they do the same thing and only have one hardcoded value:
```
purge_tags = bool(not add_only)
to_update, to_delete = compare_aws_tags(boto3_tag_list_to_ansible_dict(cur_tags.get('Tags')), tags, purge_tags)
```
And then to_delete will only be populated if tags are being removed and you don't have to reference add_only in other places.

Reply:
Ok